### DISCUSSION TITLE
How does the backup API work? Does it have limitations?

### DISCUSSION THREAD

**Discussion by fbede** [pocketbase/pocketbase#2970] [Q&A] [answered]
Hey👋,

@ganigeorgiev For sometime now, I have been wondering how the backup for pocketbase works. I've read the documentation and release notes and I didn't see any warnings, tips or caveats so I have some questions.

- How does the backup work?
- Is it a blocking task? i.e will it prevent the server from handling requests until the backup process is completed?
- Does this solution scale to large database sizes? Like 20TB? 200TB?
- How big is too big for the backup API to handle?

Please Note: I do not have an urgent need for answers, I am not running pocketbase in production yet and with my use case I hardly see the database growing more than a few gigabytes in size. I asked these questions because I feel like I could learn from how you implemented this feature.

Thank you so much for all your effort in this project.

**Comment by ganigeorgiev** [comment, marked as answer]
There are some warnings in the Admin UI when you initialize a new backup manually:
![backup_ui](https://i.imgur.com/P6ETAAb.png)

> How does the backup work?

The backup API is not anything special or complex - it creates a zip archive from your local `pb_data` (_excluding `pb_data/backups` if local backups are used_). The generated archive is either stored in `pb_data/backups` or uploaded to S3 storage, if configured.

> Is it a blocking task? i.e will it prevent the server from handling requests until the backup process is completed?

The `pb_data` zip generation is executed in a write blocking transaction, meaning that any write requests that are submitted after a backup process has started will have to wait before proceeding. Read requests are allowed as long as they are not part of a transaction to minimize the `SQLITE_BUSY` errors.

> Does this solution scale to large database sizes? Like 20TB? 200TB?

No. If you operate on such scale this type of backups usually doesn't work as the zip generation will be very slow. For such cases you may want a streaming replication, eg. something like [Litestream](https://litestream.io/) (_or even just rsync while the application is not running_).

> How big is too big for the backup API to handle?

This is difficult to say as it will depend on your use case. I've tested it on `pb_data` with 1-3 GB in size and it performed OK (_I'll consider benchmarking it more precisely in the future_), but again, if you operate on scale that requires 20TB db and disk storage I'm not even sure if PocketBase will be the correct choice in this case and definitely you'll be better with other more efficient backup options.

**Reply by fbede** [reply]
Thank you
